fix: Inconsistent Docs for Downloading ZKP Params

upbqdn opened the pull request:
## Motivation

Close #6340.

## Solution

This PR:

- adjusts the implementation so that `PARAMETER_DOWNLOAD_MAX_RETRIES` is regarded as the max number of _retries_ instead of _attempts_;
- removes the prefix `retry_` from functions that download the parameters;
- cleans up the docs a bit.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

upbqdn committed:
Use `ATTEMPTS` instead of `RETRIES`

This commit also removes the prefix `retry_` from the function names
that download the parameters. The reason for removing it was that I
consider the retries implicit, and the primary objective of the
functions is to download the params, not retry to download them.

upbqdn committed:
Docs cosmetics

arya2 commented on zebra-consensus/src/primitives/groth16/params.rs:
```diff
@@ -15,19 +15,19 @@ use crate::BoxError;
+    ///
+    /// - Zebra needs to download and cache the parameters to a shared directory, or
+    /// - Zebra checks the file hashes for both newly downloaded and previously cached files, or
+    /// - Zebra loads the parameters.
```
Nitpick/Optional:

```suggestion
    /// This static is accessed when Zebra:
    ///
    /// - needs to download and cache the parameters to a shared directory, or
    /// - checks the file hashes for both newly downloaded and previously cached files, or
    /// - loads the parameters.
```

arya2 commented on zebra-consensus/src/primitives/groth16/params.rs:
```diff
@@ -137,10 +136,10 @@ impl Groth16Parameters {
             ) {
                 retries += 1;
-                if retries >= PARAMETER_DOWNLOAD_MAX_RETRIES {
+                if retries >= PARAMETER_DOWNLOAD_MAX_ATTEMPTS {
```
Nitpick: Update variable name

```suggestion
                if num_attempts >= PARAMETER_DOWNLOAD_MAX_ATTEMPTS {
```

upbqdn replied:
Well spotted; however, I switched back to _retries_, so I'm closing this as resolved without applying the suggestion.

arya2 commented on zebra-consensus/src/primitives/groth16/params.rs:
```diff
@@ -173,23 +172,23 @@ impl Groth16Parameters {
             while let Err(error) = Groth16Parameters::download_sprout_parameters_once(sprout_path) {
                 retries += 1;
-                if retries >= PARAMETER_DOWNLOAD_MAX_RETRIES {
+                if retries >= PARAMETER_DOWNLOAD_MAX_ATTEMPTS {
```
Nitpick: Update variable name

```suggestion
                if num_attempts >= PARAMETER_DOWNLOAD_MAX_ATTEMPTS {
```

upbqdn replied:
Well spotted! However, I switched back to _retries_, so I'm resolving this without applying the suggestion.

arya2 commented on zebra-consensus/src/primitives/groth16/params.rs:
```diff
@@ -15,19 +15,19 @@ use crate::BoxError;
-/// failed for whatever reason.
-pub const PARAMETER_DOWNLOAD_MAX_RETRIES: usize = 3;
+/// Zebra will retry the download only if the previous attempt fails.
+pub const PARAMETER_DOWNLOAD_MAX_ATTEMPTS: usize = 3;
```
Optional:

Use retries and make `retries + 1` attempts.

upbqdn replied:
Done in [4208704](https://github.com/ZcashFoundation/zebra/pull/6464/commits/4208704c92c446422358baa3b89fa561473fff04).

arya2 reviewed (dismissed):
This looks great, thank you!

I may remember us saying we want to keep these as `*_RETRIES` and update the code to make `retries + 1` attempts for fixes to this finding (e.g. PR [#6460](https://github.com/ZcashFoundation/zebra/pull/6460/files#diff-1b43a508fa9f5834c283c789a7e38d9f799471988f0c2daef108c273773603f8R200)), but I'm not sure and that's not a blocker in any case.

upbqdn committed:
Apply suggestions from code review

Co-authored-by: Arya <aryasolhi@gmail.com>

upbqdn committed:
Change `ATTEMPTS` back to `RETRIES`

upbqdn commented:
> I may remember us saying we want to keep these as *_RETRIES and update the code to make retries + 1 attempts for fixes to this finding (e.g. PR [#6460](https://github.com/ZcashFoundation/zebra/pull/6460/files#diff-1b43a508fa9f5834c283c789a7e38d9f799471988f0c2daef108c273773603f8R200))

Thanks for pointing that out, I wasn't aware of that. I switched back to _retries_, adjusted the code, and edited the PR description.

arya2 commented on zebra-consensus/src/primitives/groth16/params.rs:
```diff
@@ -137,7 +136,7 @@ impl Groth16Parameters {
             ) {
                 retries += 1;
-                if retries >= PARAMETER_DOWNLOAD_MAX_RETRIES {
+                if retries > PARAMETER_DOWNLOAD_MAX_RETRIES {
```
Optional: increment retries after the if statement so that `retries` is `attempts - 1` when the original condition is checked instead of updating the condition.

upbqdn replied:
I'd say the current implementation follows the domain logic well:

- When we enter https://github.com/zcashfoundation/zebra/blob/4208704c92c446422358baa3b89fa561473fff04/zebra-consensus/src/primitives/groth16/params.rs#L186, we already know the initial attempt failed, and
- we can regard the initial attempt as the zeroth retry.
- We then count the actual retries starting at one, and
- we increment the counter after we've done a retry (not when we're about to do one), so
- we panic once we exceed the max number of retries (not when we're at the limit).

arya2 reviewed (approved):
Thank you for making the change to retries!
